### PR TITLE
Implement kotlin compiler plugin

### LINKED ISSUE #13: Implement kotlin compiler plugin for annotation processing
**Describe the problem or limitation you are having in your project:**
As we need to process annotations without a jvm target. We need a compiler plugin to process the annotations and generate code from it.

**Describe how this feature / enhancement will help you overcome this problem or limitation:**
We can do the annotation processing without a fake jvm target (see #7 ).

**Show a mock up screenshots/video or a flow diagram explaining how your proposal will work:**
None to provide

**Describe implementation detail for your proposal (in code), if possible:**
Great sources around that topic:
https://youtu.be/w-GMlaziIyo
https://youtu.be/_obNBSldffw

**If this enhancement will not be used often, can it be worked around with a few lines of code?:**
Yes if the user has to write his registrations with json.

**Is there a reason why this should be in this project and not individually solved?:**
No as it is core functionality of this project.

### PR DESCRIPTION
**Pull request by chippmann**
Resolves #13 

This pull request implements/fixes the following:
- A kotlin compiler plugin to be able to detect and process annotations in kotlin multiplatform project with an annotation processor like in normal jvm projects.
All code generation logic can be placed inside the annotation processor.

- Uses `buildSrc` and a `Dependencies` object to manage versions of all subproject.
This is done to be able to easily update versions and see all versions in one place. It also ensures that everywhere the same version of a dependency is used.

- Fixes a bug where the generated gdns files where written to the wrong directory.

- Makes the sample builds independent again as they are not part of the main project and should be independently buildable. Which in turn leads to the fact that they have to be either opened separately in intellij or added manually to the gradle tab bo be editable with ide support.

### CODE REVIEW

**Comment by piiertho** [comment]
Great job !
If I understand well, annotated elements are just logged for now ?
It seems it miss the bintray configuration to deploy artifacts and modifications to travis-ci for new artifacts.

**Comment by chippmann** [comment]
> Great job !
> If I understand well, annotated elements are just logged for now ?
> It seems it miss the bintray configuration to deploy artifacts and modifications to travis-ci for new artifacts.

Yes they are just logged for now.

You are right. I forgot those. I'll add them today.

**Comment by chippmann** [comment]
@piiertho I added the travis configurations. Please look at them carefully as I'm not sure wether they are really correct (as you know i never worked with travis before).

File: .travis/deploy.sh [hunk 1]
```diff
@@ -2,6 +2,11 @@
 if [ $3 == "linux" ]; then
   ./gradlew :tools:entry-generator:bintrayUpload -PbintrayUser=$1 -PbintrayKey=$2
   ./gradlew :tools:api-classes-generator:bintrayUpload -PbintrayUser=$1 -PbintrayKey=$2
+  ./gradlew :wrapper:annotations:bintrayUpload -PbintrayUser=$1 -PbintrayKey=$2 -Pplatform=kotlinMultiplatform
```
Comment (piiertho) [inline .travis/deploy.sh]: shouldn't it be `:tools:annotations:bintrayUpload` ?

File: .travis/deploy.sh [hunk 2]
```diff
@@ -2,6 +2,11 @@
...
+  ./gradlew :wrapper:annotations:bintrayUpload -PbintrayUser=$1 -PbintrayKey=$2 -Pplatform=metadata
+  ./gradlew :tools:annotation-processor:bintrayUpload -PbintrayUser=$1 -PbintrayKey=$2
```
Comment (piiertho) [inline .travis/deploy.sh]: Shouldn't it be `godot-annotation-processor` ?

File: tools/annotations/build.gradle.kts [hunk 3]
```diff
@@ -34,4 +37,30 @@ kotlin {
 
 tasks.build {
     finalizedBy(tasks.publishToMavenLocal)
+}
+
+val bintrayUser: String by project
+val bintrayKey: String by project
+
```
Comment (piiertho) [inline tools/annotations/build.gradle.kts]: It seems it misses the platform argument.

Comment (piiertho) [reply tools/annotations/build.gradle.kts]: also, it should only build the desired platform

**Comment by chippmann** [comment]
I rebased the branch after merge of #16 
@piiertho I'll merge it after you made the pull request to the base project

**Pull request merged**